cli/cmd/diagnose.go

```diff
@@ -0,0 +1,216 @@
+package cmd
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"context"
+	"fmt"
+	"github.com/odigos-io/odigos/cli/cmd/resources"
+	"github.com/odigos-io/odigos/cli/pkg/kube"
+	"github.com/spf13/cobra"
+	"io"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+)
+
+const (
+	logBufferSize = 1024 * 1024 // 1MB buffer size for reading logs in chunks
+)
+
+var diagnozeCmd = &cobra.Command{
+	Use:   "diagnose",
+	Short: "Diagnose Client Cluster",
+	Long:  `Diagnose Client Cluster to identify issues and resolve them. This command is useful for troubleshooting and debugging.`,
+	Run: func(cmd *cobra.Command, args []string) {
+		ctx := cmd.Context()
+		client, err := kube.CreateClient(cmd)
+		if err != nil {
+			kube.PrintClientErrorAndExit(err)
+		}
+
+		err = startDiagnose(ctx, client)
+		if err != nil {
+			fmt.Printf("The diagnose script crashed on: %v\n", err)
+		}
+	},
+}
+
+func startDiagnose(ctx context.Context, client *kube.Client) error {
+	mainTempDir, logTempDir, err := createAllDirs()
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(mainTempDir)
+
+	if err := fetchOdigosComponentsLogs(ctx, client, logTempDir); err != nil {
+		return err
+	}
+
+	if err := createTarGz(mainTempDir); err != nil {
+		return err
+	}
+	return nil
+}
+
+func createAllDirs() (string, string, error) {
+	mainTempDir, err := os.MkdirTemp("", "odigos-diagnose")
+	if err != nil {
+		return "", "", err
+	}
+
+	logTempDir := filepath.Join(mainTempDir, "Logs")
+	err = os.Mkdir(logTempDir, os.ModePerm) // os.ModePerm gives full permissions (0777)
+	if err != nil {
+		return "", "", err
+	}
+
+	return mainTempDir, logTempDir, nil
+}
+
+func fetchOdigosComponentsLogs(ctx context.Context, client *kube.Client, logDir string) error {
+	odigosNamespace, err := resources.GetOdigosNamespace(client, ctx)
+	if err != nil {
+		return err
+	}
+
+	pods, err := client.CoreV1().Pods(odigosNamespace).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	var wg sync.WaitGroup
+
+	for _, pod := range pods.Items {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			fetchPodLogs(ctx, client, odigosNamespace, pod, logDir)
+		}()
+	}
+
+	wg.Wait()
+
+	return nil
+}
+
+func fetchPodLogs(ctx context.Context, client *kube.Client, odigosNamespace string, pod v1.Pod, logDir string) {
+	for _, container := range pod.Spec.Containers {
+		logPrefix := fmt.Sprintf("Fetching logs for Pod: %s, Container: %s, Node: %s", pod.Name, container.Name, pod.Spec.NodeName)
+		fmt.Printf(logPrefix + "\n")
+
+		// Define the log file path for saving compressed logs
+		logFilePath := filepath.Join(logDir, pod.Name+"_"+container.Name+"_"+pod.Spec.NodeName+".log.gz")
+		logFile, err := os.OpenFile(logFilePath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0666)
```

Comment:
should add here
`defer logFile.Close()`
Instead of the current code which only closes the file at the happy-path where no errors occur.
This will guarantee the file is close when this function returns.

Reply:
**Why Not Use defer Inside the Loop?**
In a loop, using defer to close a file might seem like a convenient option, but deferred calls are not executed until the surrounding function returns. If you use defer inside the loop, it will accumulate all the defer calls, and the files will remain open until the function (or main program) exits. This can lead to excessive memory or resource usage, especially with a large number of iterations.

Reply:
func() {
```suggestion
		logFile, err := os.OpenFile(logFilePath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0666)
```
			defer func() {
				if err := file.Close(); err != nil {
					fmt.Printf("Error closing file %s: %v\n", fileName, err)
				}
			}()

Reply:
Ok I read a bit and this is a real cool way to solve this.. The anonymous function will be called after opening the file and because it uses defer and created inside the scope of the for-loop -> it will trigger in the end of the loop (succesfully / with error)
```
func() {
			defer func() {
				if err := file.Close(); err != nil {
					fmt.Printf("Error closing file %s: %v\n", fileName, err)
				}
			}()
```

Reply:
You can also extract the code in the loop body to a function and use a defer within that function, that way there is no risc for a long queue of defer functions.